### ISSUE TITLE
BUG: escaped closing curly brace inside MULTICHOICE not interpreted correctly

### ISSUE THREAD

**Issue by bkasel** [srobotta/moodle-tiny_cloze#21]
Hi,

I am trying out your plugin  for editing cloze questions in a multilanguage context, where I use the multilang2 filter which requires `{mlang de}` tags in answers. I was delighted to see that your plugin correctly escapes the closing curly braces 
`{1:MULTICHOICE:={mlang de\}Zylinder{mlang\}{mlang
  fr\}cylindre{mlang\}~{mlang de\}Quader{mlang\}{mlang fr\}parallélépipède rectangluaire{mlang\}}`

but when I try to edit this cloze question again with your plugin, it does not take into account the escaping correctly and breaks.

**Comment by bkasel** [comment]
Hello again,

I just noticed that you do not support the multilang2 filter, but the same problem arises when using latex formulae in a MULTICHOICE_V  question where curly braces are used all over the place.

**Comment by marcusgreen** [comment]
Could you give an example of Cloze where it uses the multilang_2 filter that this plugin can emulate

**Comment by srobotta** [comment]
Hi Ben,

I will check these issues. Two things that come up to my mind;

1. AFAIK the multilang2 filter is not applied to cloze questions, you may use the standard span tags though (I'll check this though).
2. Closing curly brackets are escaped while opening brackets are not escaped. This seems to be odd, either all are escaped or left as they are.

Do you have any other examples besides multilang tags, that are used in cloze questions and where curly brackets are used?

**Comment by bkasel** [comment]
> 1. AFAIK the multilang2 filter is not applied to cloze questions, you may use the standard span tags though (I'll check this though).

I have succesfully used multilang2 in some (not all) types of cloze questions

> 2. Closing curly brackets are escaped while opening brackets are not escaped. This seems to be odd, either all are escaped or left as they are.

I absolutely agree and I believe the source for the weirdness is how moodle parses cloze questions, the parser seems to be looking for a closing curly brace to determine where the cloze code ends, but it does not take into account any other opening curly braces.

An example not using multilang is a question containing latex code:
```
Which fraction equals 0.5? {1:MULTICHOICE_V:=\(\frac{1\}{2\}\)~\(\frac{2\}{3\}\)}
```
which renders as
![grafik](https://github.com/user-attachments/assets/0da2ba15-5d53-4f06-a205-7d0441eaa2d2)

Note that I used your plugin to create that question without escaping the closing curly braces:
![grafik](https://github.com/user-attachments/assets/f1851202-750e-49b4-928a-56394d67493a)

Then when I try to edit the question with the plugin enabled, the question breaks immeadiately.
![grafik](https://github.com/user-attachments/assets/1faff6e8-218d-4c00-8a58-391d226d1256)

**Comment by bkasel** [comment]
As for the multilang2 filter with cloze questions.... if you use the code I provided in my first post, you will get this multichoice box with only french text if your language is set to french:
![grafik](https://github.com/user-attachments/assets/f9accdb7-9493-44cf-b228-9aeb8d13dcd8)

If your language is set to english, you will see nothing because I did not specify a default version, (which is not ideal).

**Comment by srobotta** [comment]
Hi Ben,

I think I have fixed the issue (see commit in main). The closing curly braket indeed needs to be escaped while the opening braket does not. The cloze plugin just checked for the first matching closing curly braket, hence found the first whether escaped or not. I extended the regular expression to match strings until a curly closing braket without being preceeded by a backslash. This worked now for the latex example but should do for the multilang2 notation as well.

**Comment by bkasel** [comment]
Hi Stephan,

I tested it on our moodle site and it works for both latex and an multilang examples!

Thank you!

**Comment by srobotta** [comment]
Hi Ben,

Thank you for your fast confirmation. I'll close the issue.

**Comment by srobotta** [comment]
Hi Ben,

The fix is in the new 1.9 release now, that is also available in the Moodle Plugin directory.
